src/Stack/Runners.hs

```diff
@@ -169,3 +177,32 @@ withRunnerGlobal go inner = do
           | w < minTerminalWidth = minTerminalWidth
           | w > maxTerminalWidth = maxTerminalWidth
           | otherwise = w
+
+-- | Check if we should recommend upgrading Stack and, if so, recommend it.
+shouldUpgradeCheck :: RIO Config ()
+shouldUpgradeCheck = do
+  config <- ask
+  when (configRecommendUpgrade config) $ do
+    now <- getCurrentTime
+    let yesterday = addUTCTime (-24 * 60 * 60) now
+    checks <- upgradeChecksSince yesterday
+    when (checks == 0) $ do
+      mversion <- getLatestHackageVersion NoRequireHackageIndex "stack" UsePreferredVersions
+      case mversion of
+        Just (PackageIdentifierRevision _ version _) | version > mkVersion' Paths_stack.version -> do
```

Comment:
`Stack.Types.Version.stackVersion` can be used instead of `mkVersion' Paths_stack.version`.